### PR TITLE
Add V1 TaskRun Golang structs

### PR DESCRIPTION
**Pull request by JeromeJu**
# Changes

This commit adds structs and validation for v1 TaskRun, including params,
resultRef and workspaces. This is copied from v1beta1 with the following exceptions:

- omitted TaskRunResources
   - This includes the `TaskRunInputs` and `TaskRunOutputs` and respective test
- omitted `CloudEventDeliveryState`
- renamed the Go fields of `StepState.ContainerName` to `StepState.Container`and 
`SidecarState.ContainerName` to `SidecarState.Container` but not the api fields
- renamed `taskResults` to `results` 

Part of #4985
/kind misc

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [n/a] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [n/a] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

``` release-note
NONE
```

### CODE REVIEW

File: pkg/apis/pipeline/v1/taskref_types.go [hunk 1]
```diff
@@ -39,4 +39,6 @@ type TaskKind string
 const (
 	// NamespacedTaskKind indicates that the task type has a namespaced scope.
 	NamespacedTaskKind TaskKind = "Task"
+	// ClusterTaskKind indicates that task type has a cluster scope.
+	ClusterTaskKind TaskKind = "ClusterTask"
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskref_types.go:43]: ClusterTasks won't be in v1

File: pkg/apis/pipeline/v1/pod.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+import "github.com/tektoncd/pipeline/pkg/apis/pipeline/pod"
+
+// PodTemplate holds pod specific configuration
+type PodTemplate = pod.Template
+
+// MergePodTemplateWithDefault merges 2 PodTemplates together. If the same
+// field is set on both templates, the value from tpl will overwrite the value
+// from defaultTpl.
+func MergePodTemplateWithDefault(tpl, defaultTpl *PodTemplate) *PodTemplate {
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/pod.go:27]: Can any of this be pulled out of the apis package?

File: pkg/apis/pipeline/v1/taskrun_types.go [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+}
+
+// GetNamespacedName returns a k8s namespaced name that identifies this TaskRun
+func (tr *TaskRun) GetNamespacedName() types.NamespacedName {
+	return types.NamespacedName{Namespace: tr.Namespace, Name: tr.Name}
+}
+
+// IsPartOfPipeline return true if TaskRun is a part of a Pipeline.
+// It also return the name of Pipeline and PipelineRun
+func (tr *TaskRun) IsPartOfPipeline() (bool, string, string) {
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_types.go:452]: This isn't on you, but this function doesn't look like something we want to exist-- TaskRuns should in general not need to know whether they are part of a pipelinerun or not.

File: pkg/apis/pipeline/v1/taskrun_types.go [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+
+// CloudEventDeliveryState reports the state of a cloud event to be sent.
+type CloudEventDeliveryState struct {
+	// Current status
+	Condition CloudEventCondition `json:"condition,omitempty"`
+	// SentAt is the time at which the last attempt to send the event was made
+	// +optional
+	SentAt *metav1.Time `json:"sentAt,omitempty"`
+	// Error is the text of error (if any)
+	Error string `json:"message"`
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_types.go:342]: it looks like there's an API rule violation here, and I'm not sure whether it makes sense to rename the API field here or just the Go struct. Is this message expected to be present under non-error conditions or not? @afrittoli wdyt?

File: pkg/apis/pipeline/v1/taskrun_types.go [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+	if newCond != nil {
+		taskRunCondSet.Manage(trs).SetCondition(*newCond)
+	}
+}
+
+// StepState reports the results of running a step in a Task.
+type StepState struct {
+	corev1.ContainerState `json:",inline"`
+	Name                  string `json:"name,omitempty"`
+	ContainerName         string `json:"container,omitempty"`
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_types.go:300]: I think for `StepState` and `SidecarState` we can rename ContainerName -> Container to fix the API rule violation.

File: pkg/apis/pipeline/v1/taskrun_types.go [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+	// RetriesStatus contains the history of TaskRunStatus in case of a retry in order to keep record of failures.
+	// All TaskRunStatus stored in RetriesStatus will have no date within the RetriesStatus as is redundant.
+	// +optional
+	// +listType=atomic
+	RetriesStatus []TaskRunStatus `json:"retriesStatus,omitempty"`
+
+	// TaskRunResults are the list of results written out by the task's containers
+	// +optional
+	// +listType=atomic
+	TaskRunResults []TaskRunResult `json:"taskResults,omitempty"`
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_types.go:228]: We're actually going to rename this to just "results" (https://github.com/tektoncd/community/pull/775)

Comment (JeromeJu) [reply pkg/apis/pipeline/v1/taskrun_types.go]: It looks `Results` also violate the API naming rule. https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5264/pull-tekton-pipeline-build-tests/1557835200501649408

Comment (lbernick) [reply pkg/apis/pipeline/v1/taskrun_types.go]: You can rename both the field name and the api field (in the JSON tag) to "results"

File: pkg/apis/pipeline/v1/taskrun_types.go (hunk 3 above)
Comment (JeromeJu) [reply pkg/apis/pipeline/v1/taskrun_types.go:452]: Thanks Lee. Opened [#5320](https://github.com/tektoncd/pipeline/issues/5320) to track this down.

File: pkg/apis/pipeline/v1/pod.go (hunk 2 above)
Comment (JeromeJu) [reply pkg/apis/pipeline/v1/pod.go:27]: Can I ask for hint where do we put this? I tried to put under the pkg/pod package but that would introduce cyclic import.

File: pkg/apis/pipeline/v1/taskrun_types.go (hunk 4 above)
Comment (afrittoli) [reply pkg/apis/pipeline/v1/taskrun_types.go:342]: `CloudEventDeliveryState` is only used for the `CloudEventPipelineResource` so given the fact that we are not having `PipelineResources` in `v1` both the API and structs should go away in `v1`.

Comment (JeromeJu) [reply pkg/apis/pipeline/v1/taskrun_types.go]: Thanks @afrittoli. Removed both from v1.

File: pkg/apis/pipeline/v1/pod.go (hunk 2 above)
Comment (lbernick) [reply pkg/apis/pipeline/v1/pod.go:27]: it can probably go in pkg/apis/pipeline/pod

File: pkg/apis/pipeline/v1/taskrun_types.go (hunk 4 above)
Comment (lbernick) [reply pkg/apis/pipeline/v1/taskrun_types.go]: Could you please update the commit message and PR description to reflect that these structs have been removed, that you renamed the StepState.ContainerName and SidecarState.ContainerName Go fields (but not the api fields), and that the "taskResults" api field was renamed to "results"

Comment (JeromeJu) [reply pkg/apis/pipeline/v1/taskrun_types.go]: Updated. Thanks Lee @lbernick  for the summary.

File: pkg/apis/pipeline/v1/pod.go (hunk 2 above)
Comment (JeromeJu) [reply pkg/apis/pipeline/v1/pod.go:27]: Thanks Lee. Added #5329

File: pkg/apis/pipeline/v1/taskrun_validation.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+		errs = errs.Also(ts.TaskRef.Validate(ctx).ViaField("taskRef"))
+	}
+	// Validate TaskSpec if it's present.
+	if ts.TaskSpec != nil {
+		errs = errs.Also(ts.TaskSpec.Validate(ctx).ViaField("taskSpec"))
+	}
+
+	errs = errs.Also(ValidateParameters(ctx, ts.Params).ViaField("params"))
+	errs = errs.Also(ValidateWorkspaceBindings(ctx, ts.Workspaces).ViaField("workspaces"))
+	if ts.Debug != nil {
```
Comment (JeromeJu) [inline pkg/apis/pipeline/v1/taskrun_validation.go:63]: Shall I sync with 5291 here for the [Resources related code](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/taskrun_validation.go/#L68)

Comment (lbernick) [reply pkg/apis/pipeline/v1/taskrun_validation.go:63]: I'm not exactly sure what you mean here-- we shouldn't have resources in v1 so no need to validate them

Comment (JeromeJu) [reply pkg/apis/pipeline/v1/taskrun_validation.go:68]: Sorry this was mistaken with some other changes when I looked at the change log.

File: pkg/apis/pipeline/v1/taskrun_types.go [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+	// Steps describes the state of each build step container.
+	// +optional
+	// +listType=atomic
+	Steps []StepState `json:"steps,omitempty"`
+
+	// CloudEvents describe the state of each cloud event requested via a
+	// CloudEventResource.
+	// +optional
+	// +listType=atomic
+	CloudEvents []CloudEventDelivery `json:"cloudEvents,omitempty"`
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_types.go:218]: I don't think this is necessary based on @afrittoli's comment about this being used only for PipelineResources (which matches up with the comment here that it's used for CloudEventResource)

File: pkg/apis/pipeline/v1/taskrun_types.go [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+const (
+	// CloudEventConditionUnknown means that the condition for the event to be
+	// triggered was not met yet, or we don't know the state yet.
+	CloudEventConditionUnknown CloudEventCondition = "Unknown"
+	// CloudEventConditionSent means that the event was sent successfully
+	CloudEventConditionSent CloudEventCondition = "Sent"
+	// CloudEventConditionFailed means that there was one or more attempts to
+	// send the event, and none was successful so far.
+	CloudEventConditionFailed CloudEventCondition = "Failed"
+)
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_types.go:326]: this can be removed

File: pkg/apis/pipeline/v1/taskrun_types_test.go [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+	v1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1"
+	"github.com/tektoncd/pipeline/test/diff"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+)
+
+// TODO: issues#5111 related, shall be in v1_test
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_types_test.go:35]: this can be removed

**Review by lbernick — APPROVED** [review]

File: pkg/apis/pipeline/v1/taskrun_types.go (hunk 8 above)
Comment (JeromeJu) [reply pkg/apis/pipeline/v1/taskrun_types.go]: Thanks Lee. Now removed. Previously only deleted the one related with CloudEventDeliveryStatus.

**Review by jerop — COMMENTED** [review]
/retest

**Comment by dibyom** [comment]
This LGTM. Can you squash up the commits?

**Comment by JeromeJu** [comment]
Thanks @dibyom, squashed and retested.

File: pkg/apis/pipeline/v1/taskrun_validation_test.go [hunk 11]
```diff
@@ -0,0 +1,887 @@
...
+	corev1 "k8s.io/api/core/v1"
+	corev1resources "k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+)
+
+func TestTaskRun_Invalidate(t *testing.T) {
+	tests := []struct {
+		name    string
+		taskRun *v1beta1.TaskRun
```
Comment (chitrangpatel) [inline pkg/apis/pipeline/v1/taskrun_validation_test.go:39]: @JeromeJu Shouldn't this be `v1.TaskRun`? I noticed it while I was trying to rebase with some of my changes.

Comment (lbernick) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:39]: Thanks Chitrang for spotting, yes this should be v1.TaskRun

Comment (JeromeJu) [reply pkg/apis/pipeline/v1/taskrun_validation_test.go:39]: Thanks @chitrangpatel. Working on this.

**Pull request merged**